azure-iot-hub/tests/test_digital_twin_client.py

```diff
@@ -107,6 +112,63 @@ def test_instantiates_with_connection_string_no_shared_access_key(self):
             DigitalTwinClient(connection_string)
 
 
+@pytest.mark.describe("DigitalTwinClient - .from_connection_string()")
+class TestFromConnectionString:
+    @pytest.mark.it(
+        "Creates an instance of ConnectionStringAuthentication and passes it to IotHubGatewayServiceAPIs constructor"
+    )
+    def test_connection_string_auth(self, mocker):
+        connection_string_auth_init_mock = mocker.patch.object(
+            azure.iot.hub.auth, "ConnectionStringAuthentication"
+        )
+        connection_string_auth_mock = connection_string_auth_init_mock.return_value
+        connection_string_auth_mock.__getitem__.return_value = fake_hostname
+        protocol_client_init_mock = mocker.patch.object(
+            azure.iot.hub.protocol.iot_hub_gateway_service_ap_is, "IotHubGatewayServiceAPIs"
+        )
+
+        connection_string = "HostName={hostname};DeviceId={device_id};SharedAccessKeyName={skn};SharedAccessKey={sk}".format(
+            hostname=fake_hostname,
+            device_id=fake_device_id,
+            skn=fake_shared_access_key_name,
+            sk=fake_shared_access_key,
+        )
+
+        DigitalTwinClient.from_connection_string(connection_string=connection_string)
+
+        assert connection_string_auth_init_mock.call_args == mocker.call(connection_string)
+        assert protocol_client_init_mock.call_args == mocker.call(
+            connection_string_auth_mock, "https://" + connection_string_auth_mock["HostName"]
+        )
+
+
+@pytest.mark.describe("DigitalTwinClient - .from_token_credential()")
+class TestFromTokenCredential:
+    @pytest.mark.it(
+        "Creates an instance of AzureIdentityCredentialAdapter and passes it to IotHubGatewayServiceAPIs constructor"
+    )
+    def test_token_credential_auth(self, mocker):
+        azure_identity_credential_adapter_init_mock = mocker.patch.object(
+            azure.iot.hub.auth, "AzureIdentityCredentialAdapter"
+        )
+        azure_identity_credential_adapter_mock = (
+            azure_identity_credential_adapter_init_mock.return_value
+        )
+        mock_azure_identity_TokenCredential = mocker.Mock()
+        protocol_client_init_mock = mocker.patch.object(
+            azure.iot.hub.protocol.iot_hub_gateway_service_ap_is, "IotHubGatewayServiceAPIs"
+        )
+
+        DigitalTwinClient.from_token_credential(fake_hostname, mock_azure_identity_TokenCredential)
+
+        assert azure_identity_credential_adapter_init_mock.call_args == mocker.call(
+            mock_azure_identity_TokenCredential
+        )
+        assert protocol_client_init_mock.call_args == mocker.call(
```

Comment:
I had said above that perhaps you could avoid mocking this out, but if it turns out you can't, you may want to consider creating a fixture that will always mock it out so you don't have to do it manually.